#!/usr/bin/python
# -*- coding:utf8 -*-

import paho.mqtt.client as mqtt
from collections import OrderedDict
import json
import rospy
import random
import time
from threading import *
from std_msgs.msg import Int32, Float32, Float32MultiArray

class AN_Mqtt():
    def __init__(self):
        self.mqtt_host = "192.168.10.124"
        self.mqtt_port = 50002
        self.mqttClient = mqtt.Client()
        self.AN_Mission_Msg = 0
        #self.start_mqtt()

    # è¿æ¥MQTTæœåŠ¡å™?
    def on_mqtt_connect(self):
        self.mqttClient.connect(self.mqtt_host, self.mqtt_port, 60)

    # subscribe æ¶ˆæ¯
    def on_subscribe(self):
        self.mqttClient.subscribe("/HG_DEV/ZK_ALL", 2)
        self.mqttClient.on_message = self.on_message_come  # æ¶ˆæ¯åˆ°æ¥å¤„ç†å‡½æ•°
    #   å?åŠ¨mqttè¿æ¥
    def start_mqtt(self):
        self.on_mqtt_connect()
        self.on_subscribe()
        self.mqttClient.loop_start()

    # publish æ¶ˆæ¯
    def on_publish(self, payload, topic="/HG_DEV/ZK_ALL_REQ", qos=2):
        print("pub msg: %s to %s" % (payload, topic))
        self.mqttClient.publish(topic, payload, qos)

    # æ¶ˆæ¯å¤„ç†å‡½æ•°
    def on_message_come(self, lient, userdata, msg):
        print("from: " + msg.topic + " " + ":" + msg.payload)
        sub_msg = json.loads(msg.payload.encode('utf-8'))
        if sub_msg["name"] == "ZK" and sub_msg["dir"] == "AD" and sub_msg["mission"] == 1:
            self.AN_Mission_Msg = 1

class AN_Mission_Test:
    def __init__(self):
        rospy.init_node("an_node")
        self.host_ip = rospy.get_param("~host_ip", "192.168.10.124")
        self.host_port = rospy.get_param("~host_port", "50002")
        # è§†è?‰è¯†åˆ?æ¨¡å¼,0:AGVä¸Šæ²¡æœ‰ç ,ç›²æ”¾; 1:AGVä¸Šæœ‰ç ?,è¯†åˆ«æ”¾ç½®
        self.vision_mode = rospy.get_param("~vision_mode", "0")

        # æŠ“å–è¯†åˆ«ç‚?
        self.grasp_x1 = int(rospy.get_param("~grasp_x1", "80"))
        self.grasp_y1 = int(rospy.get_param("~grasp_y1", "490"))
        self.grasp_z1 = int(rospy.get_param("~grasp_z1", "10"))
        
        # æ”¾ç½®è¯†åˆ«ç‚?
        self.push_x1 = int(rospy.get_param("~push_x1", "200"))
        self.push_y1 = int(rospy.get_param("~push_y1", "60"))
        self.push_z1 = int(rospy.get_param("~push_z1", "30"))

        # é»˜è?¤æ”¾ç½?ç‚?
        self.real_push_x1 = int(rospy.get_param("~real_push_x1", "200"))
        self.real_push_y1 = int(rospy.get_param("~real_push_y1", "60"))
        self.real_push_z1 = int(rospy.get_param("~real_push_z1", "250"))


        # æŠ“å–è¯†åˆ«ç‚?
        self.grasp_x2 = int(rospy.get_param("~grasp_x2", "80"))
        self.grasp_y2 = int(rospy.get_param("~grasp_y2", "490"))
        self.grasp_z2 = int(rospy.get_param("~grasp_z2", "10"))
        
        # æ”¾ç½®è¯†åˆ«ç‚?
        self.push_x2 = int(rospy.get_param("~push_x2", "200"))
        self.push_y2 = int(rospy.get_param("~push_y2", "60"))
        self.push_z2 = int(rospy.get_param("~push_z2", "30"))

        # é»˜è?¤æ”¾ç½?ç‚?
        self.real_push_x2 = int(rospy.get_param("~real_push_x2", "200"))
        self.real_push_y2 = int(rospy.get_param("~real_push_y2", "60"))
        self.real_push_z2 = int(rospy.get_param("~real_push_z2", "250"))

        self.an_req_msg = OrderedDict()
        self.an_req_msg["name"] = "AD"
        self.an_req_msg["dir"] = "ZK"
        self.an_req_msg["mission"] = 1
        self.an_req_msg["state"] = "working"
        self.an_req_json = json.dumps(self.an_req_msg)

        self.send_time = 0
        self.wait_time = 0
        self.an_current_x = 0.0
        self.an_current_y = 0.0
        self.an_current_z = 0.0
        self.an_grasp_state = 0
        self.an_target_pose = [0.0, 0.0, 0.0]
        # æ”¾ç½®ç‚?
        self.push_target_pose = [0.0, 0.0, 0.0]
        self.an_control = rospy.Publisher('/Pall_POS_SET', Float32MultiArray, queue_size=5) # å‘å¸ƒåæ ‡æ§åˆ¶è¯é??
        self.an_control_pos = rospy.Subscriber("/Pall_CURR_POS", Float32MultiArray, self.get_an_pose)   # è®¢é˜…è®¾å?‡å½“å‰xyzåæ ‡
        self.an_grasp_control = rospy.Publisher('/Pall_Grasp_Topic', Int32, queue_size=5)   # æ§åˆ¶æœ?ç«?çˆ?ï¼?1ï¼šæŠ“å–ï¼›0ï¼šæ¾å¼€ï¼?
        self.an_grasp_state = rospy.Subscriber("/Pall_GRAB_STATUS", Int32, self.get_an_control) #   è·å–çˆ?çš„çŠ¶æ€?

        # ä¸»æ§è§†è?‰è¯†åˆ?è¯é??
        rospy.Subscriber('target_pose', Float32MultiArray, self.get_target_pose)
        rospy.Subscriber('push_target_pose', Float32MultiArray, self.get_push_data)
        self.set_push_state = rospy.Publisher('/get_push_pose', Int32, queue_size=5)

        self.AN_Mqtt = AN_Mqtt()
        self.AN_Mqtt.mqtt_host = self.host_ip
        self.AN_Mqtt.mqtt_port = self.host_port
        self.AN_Mqtt.start_mqtt()

        self.th = Thread(target=self.get_mission)
        self.th.start()
    
    def get_target_pose(self, data):
        self.an_target_pose = data.data
        print(data.data)
    
    def get_push_data(self, data):
        self.push_target_pose = data.data
        print(data.data)

    def get_mission(self):
        while True:
            if self.AN_Mqtt.AN_Mission_Msg == 1:
                self.AN_Mqtt.AN_Mission_Msg = 0
                self.an_req_msg["state"] = "working"
                self.an_req_json = json.dumps(self.an_req_msg)
                self.AN_Mqtt.on_publish(self.an_req_json, "/HG_DEV/ZK_ALL_REQ", 2)
                time.sleep(0.5)
                print("å¼€å§‹ä¸Šè´?")
                # å¦‚æœAGVä¸Šä½¿ç”¨è¯†åˆ?æ”¾ç½®ï¼ˆAGVä¸Šé¢è´´ä¸Šè¯†åˆ«ï¼?,å…ˆåˆ°æ”¾ç½®ç‚¹è¯†åˆ?ç‚¹è¯†åˆ?
                if self.vision_mode == 1:
                    # å‘å¸ƒè·å–æ”¾ç½®è¯†åˆ«ç‚¹ä¿¡æ?
                    self.set_push_state.publish(1)
                    # ç§»åŠ¨åˆ°æ”¾ç½?è¯†åˆ«ç‚?
                    if self.set_an_pose([self.push_x, self.push_y, self.push_z]):
                        # è¯†åˆ«æ”¾ç½®åæ ‡å¹¶è?°å½•(æ™?é€šarç è¯†åˆ?)
                        try:
                            data = rospy.wait_for_message("push_target_pose", Float32MultiArray)
                            self.push_target_pose = data.data
                        except:
                            pass

                # ç§»åŠ¨åˆ°æŠ“å–è¯†åˆ?ç‚?
                if self.set_an_pose([self.grasp_x1, self.grasp_y1, self.grasp_z1]):
                    try:
                        # è¯†åˆ«æŠ“å–åæ ‡å¹¶æ‰§è¡?(ä»ä¸»æ§ç•Œé¢å‘é€è?†è?‰è¯†åˆ?ç‰©ä½“ç¼–å·,å¹¶æ¥å—åæ ‡è¯é¢?)
                        data = rospy.wait_for_message("target_pose", Float32MultiArray)
                        if self.set_an_pose(data.data):
                            #time.sleep(1.5)
                            self.set_an_grasp_control(1)
                        #time.sleep(1.0)
                        #while True:
                            #if self.an_grasp_state:  # å·²ç»æŠ“ä¸Šæ?
                            #    break
                            #else:
                                #self.set_an_grasp_control(1)
                                #time.sleep(1.0)
                        # å¾€ä¸Šæèµ·ç‰©ä½“å¹¶;ç§»åŠ¨åˆ°æ”¾ç½?è¯†åˆ«ç‚?
                        if self.set_an_pose([0.0, 0.0, 30.0]):
                            if self.set_an_pose([self.push_x1, self.push_y1, self.push_z1]):
                                if self.vision_mode == 1 and self.push_target_pose[2]!=0.0:
                                    if self.set_an_pose(self.push_target_pose): # è¯†åˆ«æ”¾ç½®ç‚?
                                        print("push target")
                                else:
                                    if self.set_an_pose([self.real_push_x1, self.real_push_y1, self.real_push_z1]): # ç›²æ”¾ä½ç½®ç‚?
                                        print("push target")
                                time.sleep(1.0)
                                self.set_an_grasp_control(0)
                                time.sleep(1.0)
                                while True:
                                    if not self.an_grasp_state:  # å·²ç»æ”¾å¼€
                                        break
                                    else:
                                        self.set_an_grasp_control(0)

                                if self.set_an_pose([0.0, 0.0, 20.0]):  # æ”¾ä¸Šå»åå…ˆå¾€ä¸?
                                    try:
                                        # è¯†åˆ«æŠ“å–åæ ‡å¹¶æ‰§è¡?(ä»ä¸»æ§ç•Œé¢å‘é€è?†è?‰è¯†åˆ?ç‰©ä½“ç¼–å·,å¹¶æ¥å—åæ ‡è¯é¢?)
                                        data1 = rospy.wait_for_message("target_pose1", Float32MultiArray)
                                        if self.set_an_pose(data1.data):
                                            #time.sleep(1.5)
                                            self.set_an_grasp_control(1)
                                        #time.sleep(1.0)
                                        #while True:
                                            #if self.an_grasp_state:  # å·²ç»æŠ“ä¸Šæ?
                                            #    break
                                            #else:
                                                #self.set_an_grasp_control(1)
                                                #time.sleep(1.0)
                                        # å¾€ä¸Šæèµ·ç‰©ä½“å¹¶;ç§»åŠ¨åˆ°æ”¾ç½?è¯†åˆ«ç‚?
                                        if self.set_an_pose([0.0, 0.0, 30.0]):
                                            if self.set_an_pose([self.push_x2, self.push_y2, self.push_z2]):
                                                if self.vision_mode == 1 and self.push_target_pose[2]!=0.0:
                                                    if self.set_an_pose(self.push_target_pose): # è¯†åˆ«æ”¾ç½®ç‚?
                                                        print("push target")
                                                else:
                                                    if self.set_an_pose([self.real_push_x2, self.real_push_y2, self.real_push_z2]): # ç›²æ”¾ä½ç½®ç‚?
                                                        print("push target")
                                                time.sleep(1.0)
                                                self.set_an_grasp_control(0)
                                                time.sleep(1.0)
                                                while True:
                                                    if not self.an_grasp_state:  # å·²ç»æ”¾å¼€
                                                        break
                                                    else:
                                                        self.set_an_grasp_control(0)

                                                if self.set_an_pose([0.0, 0.0, 20.0]):  # æ”¾ä¸Šå»åå…ˆå¾€ä¸?
                                                    
                                                    self.an_req_msg["state"] = "finish"
                                                    self.an_req_json = json.dumps(self.an_req_msg)
                                                    self.AN_Mqtt.on_publish(self.an_req_json, "/HG_DEV/ZK_ALL_REQ", 2)
                                    except:
                                        pass

                    except:
                        pass
            
                # if self.set_an_pose([150.0, 650.0, 510.0]):  # æŠ“å–ç‚?
                #     self.set_an_grasp_control(1)
                # while True:
                #     if self.an_grasp_state:  # å·²ç»æŠ“ä¸Šæ?
                #         break

                # if self.set_an_pose([0.0, 0.0, 300.0]):  # æŠ“åˆ°åå…ˆå¾€ä¸?
                #     if self.set_an_pose([130.0, 75.0, 520.0]):  # æ”¾ç½®ç‚?
                #         self.set_an_grasp_control(0)
                #     while True:
                #         if not self.an_grasp_state:  # å·²ç»æ”¾ä¸Šå?
                #             if self.set_an_pose([0.0, 0.0, 100.0]):  # æ”¾ä¸Šå»åå…ˆå¾€ä¸?
                #                 break

    #   è·å–å½“å‰ä½å§¿
    def get_an_pose(self, data):
        self.an_current_x = data.data[0]
        self.an_current_y = data.data[1]
        self.an_current_z = data.data[2]

    #  æ§åˆ¶èµ? x y z 
    def set_an_pose(self, data):
        if data[0] != 0.0 or data[1] != 0.0:  # x,yä¸ä¸º0,å…ˆç§»åŠ¨x,yå†ç§»åŠ¨z
            move_xy = [data[0], data[1], self.an_current_z]
            target_pose = Float32MultiArray()
            target_pose.data = move_xy
            print(target_pose)
            for i in range(4):
                self.an_control.publish(target_pose)
            while True:
                if abs(self.an_current_x - data[0]) < 5 and abs(self.an_current_y - data[1]) < 5:
                    break
            move_z = [data[0], data[1], data[2]]
            target_pose = Float32MultiArray(data=move_z)
            print(target_pose)
            for i in range(4):
                self.an_control.publish(target_pose)
            while True:
                if (abs(self.an_current_z - data[2]) < 30):
                    self.wait_time = 0
                    break
                else:
                    self.wait_time += 1
                    time.sleep(1.0)
                if self.wait_time > 15:
                    self.wait_time = 0
                    break
            return True
        else:  # x,yä¸?0åˆ™å•ç‹?æ§åˆ¶z
            move_z = [self.an_current_x, self.an_current_y, data[2]]
            target_pose = Float32MultiArray(data=move_z)
            for i in range(4):
                self.an_control.publish(target_pose)
            while True:
                if (abs(self.an_current_z - data[2]) < 30):
                    time.sleep(2.0)
                    break
            return True

    #   è·å–çˆ?çš„çŠ¶æ€?
    def get_an_control(self, data):
        self.an_grasp_state = data.data

    def set_an_grasp_control(self, data):
        #for i in range(3):
        self.an_grasp_control.publish(data)
        time.sleep(0.2)

if __name__ == '__main__':
    an_start = AN_Mission_Test()
    rospy.spin()
